### DISCUSSION TITLE
Does restoring cached outputs first delete the existing files?

### DISCUSSION THREAD

**Discussion by bryanjswift** [moonrepo/moon#546] [Q&A] [answered]
By way of context setting I'll describe what has led me to this question. I've been working on converting a large-ish monorepo to use moon rather than makefiles. I'm fairly near the end of the conversion and have started to test the developer experience locally. When developing locally there are typically three different HTTP based services running locally, each of the local HTTP servers have some hot reloading or restart on changes behavior. During my testing I have noticed that sometimes if I rebuild the project while the local HTTP server processes are running some cached files will be restored such that I only get about half of the file contents. A different local HTTP server process will die completely with a stack trace indicating it failed in the `tryReadFile` function of the TypeScript library during compilation.

Basically, I don't have enough information to create a good bug report but want to understand the behavior of moon when restoring cached outputs a bit better.

**Comment by milesj** [comment, marked as answer]
@bryanjswift Great question. 

If no `outputs` is defined, then no hydration occurs, and the state of the project folder just stays as-is.

If there are `outputs`, then we create the tarballs in `.moon/cache/outputs`. When a task is ran, we generate a hash, and if there's already an existing output with that hash, then we hydrate it. Hydration basically follows these steps:

- For each file in the tarball, unpack it to the output destination if its a new file or the file contents are different than an existing file. We check file size and the contents.
- For files that were already in the output destination, and not in the tarball, we assume these are stale files from a previous build, and we delete them.

So yes, to answer your question, we do delete files, but we don't delete _all files_, just those that shouldn't be around anymore. This is all handled via our `TreeDiffer`: https://github.com/moonrepo/moon/blob/master/crates/core/archive/src/tree_differ.rs

Can you paste your tasks config? May help uncover any issues.

**Reply by bryanjswift** [reply]
The task I have that is restoring the latter half of the file sometimes is this one. `packages/site-builder/src/generated/admin-types.ts` is the file that is intermittently restoring badly. The `admin-types.ts` file is generated via https://the-guild.dev/graphql/codegen. I would expect the file to compare exactly, in contents and hash, but I haven't done the testing or tried to set up a minimal test case to reproduce. The full tasks config is spread over a lot of files in the monorepo. The `messaging-api:gen-admin-schema` and `messaging-api:gen-attendee-schema` dependencies generate files I just realized should be in the inputs.

```yml
  gen-gql-types:
    command: 'graphql-codegen --silent'
    deps:
      - 'messaging-api:gen-admin-schema'
      - 'messaging-api:gen-attendee-schema'
    inputs:
      - '/codegen.yml'
      - '/component-packages/access-code/src/**/*.ts'
      - '/component-packages/attendee-container/src/**/*.ts'
      - '/component-packages/video/src/**/*.ts'
      - '/component-packages/rtc-room-container/src/**/*.ts'
      - '/packages/admin-gql-types/src/**/*.ts'
      - '/packages/attendee-ui-types/src/**/*.ts'
      - '/packages/site-builder/src/**/*.ts'
      - '/packages/site-builder/src/**/*.tsx'
    outputs:
      - 'component-packages/access-code/src/generated/attendee-types.ts'
      - 'component-packages/attendee-container/src/generated/attendee-types.ts'
      - 'component-packages/rtc-room-container/src/generated/attendee-types.ts'
      - 'component-packages/video/src/generated/attendee-types.ts'
      - 'packages/admin-gql-types/src/generated/admin-types.ts'
      - 'packages/attendee-ui-types/src/generated/attendee-types.ts'
      - 'packages/site-builder/src/generated/admin-types.ts'
```

**Reply by milesj** [reply]
Yeah, if there are inputs that may alter what the outputs look like, they'll definitely need to be listed.

Another option here is to simply not use `outputs`. Outputs are useful for caching long builds, like a webpack application build. If the GQL codegen is fast and only takes like a second, it may be best to not cache it.

**Reply by bryanjswift** [reply]
I'm fairly sure that the missing inputs isn't the culprit but I agree they should be listed. I experienced identical behavior after adding the missing `inputs`. I've also found outputs useful for caching medium long build steps. The cache restoring is much faster than a second.

**Reply by bryanjswift** [reply]
Additionally, if I don't store `outputs` then the running a build on a clean project doesn't restore these files so then I would explicitly need to set `cache` to `false` to ensure the task _always_ runs rather than only when it needs to run.

**Reply by milesj** [reply]
I wish there was an easy way to debug this, but you'll most likely need to compare the hash files in `.moon/cache/hashes` to figure out what's triggering a new build, and then attempt to deduce which file is either missing from the output tarball, or is the wrong value.

It's also possible that when the output was created the project was in a bad state, so when it rehydrates with that output, it goes back into that state.

**Reply by bryanjswift** [reply]
I was able to address my immediate problem by refining the list of `inputs`, by which I mean I eliminated some input files that could not actually change the outputs which for now causes the task to not re-run as often. I also think I misattributed the cause of my error. I've marked your original post as the answer because it answered the question I initially posed. Thanks for your time helping me to understand some more about how moon works internally.

**Reply by milesj** [reply]
@bryanjswift Awesome, glad you got it working!
